src/ObjectPool/src/DefaultObjectPool.cs

```diff
@@ -39,66 +38,62 @@ public DefaultObjectPool(IPooledObjectPolicy<T> policy)
     /// <param name="maximumRetained">The maximum number of objects to retain in the pool.</param>
     public DefaultObjectPool(IPooledObjectPolicy<T> policy, int maximumRetained)
     {
-        _policy = policy ?? throw new ArgumentNullException(nameof(policy));
-        _fastPolicy = policy as PooledObjectPolicy<T>;
-        _isDefaultPolicy = IsDefaultPolicy();
-
-        // -1 due to _firstItem
-        _items = new ObjectWrapper[maximumRetained - 1];
-
-        bool IsDefaultPolicy()
-        {
-            var type = policy.GetType();
-
-            return type.IsGenericType && type.GetGenericTypeDefinition() == typeof(DefaultPooledObjectPolicy<>);
-        }
+        // cache the target interface methods, to avoid interface lookup overhead
+        _createFunc = policy.Create;
+        _returnFunc = policy.Return;
+        _maxCapacity = maximumRetained - 1;  // -1 to account for _fastItem
     }
 
     /// <inheritdoc />
     public override T Get()
     {
-        var item = _firstItem;
-        if (item == null || Interlocked.CompareExchange(ref _firstItem, null, item) != item)
+        var item = _fastItem;
+        if (item == null || Interlocked.CompareExchange(ref _fastItem, null, item) != item)
         {
-            var items = _items;
-            for (var i = 0; i < items.Length; i++)
+            if (_items.TryDequeue(out item))
             {
-                item = items[i].Element;
-                if (item != null && Interlocked.CompareExchange(ref items[i].Element, null, item) == item)
-                {
-                    return item;
-                }
+                _ = Interlocked.Decrement(ref _numItems);
```

Comment:
nit: We don't use this style
```suggestion
                Interlocked.Decrement(ref _numItems);
```